### PR TITLE
docs(klip): klip-7: Kafka Connect Integration

### PR DESCRIPTION
**Pull request by agavra**
# KLIP 7 - Kafka Connect Integration

**Author**: agavra | 
**Release Target**: 5.4 | 
**Status**: _In Discussion_ | 
**Discussion**: link

**tl;dr:** _provide first-class integration with Kafka connect for KSQL data ingress and egress_
           
## Motivation and background

Databases provide three functions:
- Data ingress (i.e. inserting data into the database)
- Data transform (i.e. deriving new data from old)
- Data egress (i.e. querying and/or ETL)

In order to develop KSQL as the streaming database, we must provide a complete story around each of
these functionality. Today, KSQL is a compelling tool to accomplish data transformations but relies
on external mechanisms to import and export data. This KLIP lays out foundations for an integration 
with Kafka Connect, a system that provides powerful primitives for data ingress and egress for 
Apache Kafka.

### Motivating Example

This document will extend the canonical KSQL example of joining pageviews with a user table, but in
this document we will assume that the user table is stored in a JDBC compatible database, as opposed
to a Kafka changelog topic. After implementing this KLIP, it should be possible for a KSQL user to
perform a stream-table join on these two without leaving KSQL's console.

## Scope

This KLIP will:
- define a SQL-like language for basic connector management
- provide a guiding principles and high level architecture for a connect-ksql integration

This KLIP will _not_:
- propose improvements to Kafka Connect such as easing the configuration burden, but will discuss
  how to leverage future Connect improvements

## Value/Return

This proposal targets _usability_ improvements for KSQL and simplifies bootstrapping KSQL for users
that may already have some, but not all, of their data in Kafka. As with the motivating example,
KSQL users will be able to easily model external systems within KSQL. The target user has already
played around with KSQL and is working on prototyping their first real application with existing
data.

### Architecture

There are three ways to integrate KSQL with Connect:
1. Run them separately and supply syntax to communicate between them
1. Embed Connect into KSQL, managing individual connect workers as part of the KSQL ecosystem
1. Package Connect and KSQL into a single JVM, but just spin them up at the top level

There is some consensus that at scale KSQL and Connect should run as separate components. This
means that the API that we design will not be able to leverage any "embedded-ness" of Connect,
nullifying most of the benefits of the second approach. 

There are still, however, benefits to packaging connect and KSQL in the same JVM for a seamless
experience for new users. Namely, new users will only need to download a single package, spin up
a single process and maintain only a single application. There are some (solvable) concerns with
this approach, but for the reason outlined above, the syntax will be developed with the assumption
that Connect is running as a separate component.

**The recommendation is to support 1 & 3, tabling 2 unless a need for it comes in the future**

### Syntax

Analogous to DDL and DML, the Connect syntax will separate actions that have side effects with
actions that simply declare metadata on top of existing data. This solves two issues:

- A single connector can have many topics, which makes modeling them as a single statement difficult
- A connector can be modeled in different ways (e.g. stream/table) and connector lifecycle should
  be independent with how KSQL interprets it

There are a few more minor points of consideration:
- All "DML" commands that cause changes in connect will not be submitted to the command topic;
  instead they will be issued on the REST server (this follows the pattern that all external-state
  modifications are not distributed)
- Some connectors will have KSQL-specific wrapper implementations that will allow simpler 
  configuration and tighter integration. These connectors will also automatically issue the DDL 
  statements for any topics that they create, when possible.

**Sample Ingress**

The first step is to create and start a Connector via the REST API (`POST /connectors`)
```sql
ksql> CREATE EXTERNAL SOURCE my-postgres-jdbc WITH ( \
      connector.class="io.confluent.connect.jdbc.jdbcSourceConnector", \ 
      tasks.max=1, \
      connection.url=jdbc:postgresql://localhost:5432/my-db, \
      ...);

--------------
SOURCE CREATED
--------------
```

If the connector is one of the KSQL supported connectors, we can have default configuration options
and tighter integration specified by `CREATE EXTERNAL SOURCE(type='JDBC') WITH (...);`.

*Note: as Connect manages to simplify the configuration, whether through templates or any other
means, we will immediately be able to leverage this simplification in KSQL's WITH clause.*

Next, we provide tools to understand the status of the connector and corresponding topics hitting
the `GET /connectors` and `GET /connectors/<name>` REST endpoints:
```sql
ksql> SHOW EXTERNAL SOURCES;

    name         |              connector                        |   topic prefix
-----------------|-----------------------------------------------|--------------------
my-postgres-jdbc | io.confluent.connect.jdbc.JdbcSourceConnector | test-postgres-jdbc

ksql> DESCRIBE my-postgres-jdbc;

name: my-postgres-jdbc
connector: io.confluent.connect.jdbc.SourceConnector
topic-prefix: test-postgres-jdbc

   worder id     | task | status
-----------------|------|---------
10.200.7.69:8083 | 0    | RUNNING

 name      |     topic                     |  schema
-----------|-------------------------------|--------------------------------------
user-table | test-postgres-jdbc-user-table | id VARCHAR, user VARCHAR, age INTEGER
page-table | test-postgres-jdbc-page-table | id VARCHAR, page VARCHAR, url VARCHAR
```

Then the user can create streams or tables from the underlying topics (this can happen automatically
for some set of blessed connectors):
```sql
ksql> CREATE TABLE users_original WITH (kafka_topic = 'test-postgres-jdbc-user-table', ...);

--------------
STREAM CREATED
--------------
```

At this point, assuming the `pageviews_original` stream is already created, we can join the newly
created `users_original` table with the `pageviews_original` stream. Finally, the user can drop and 
terminate the connectors.
```sql
ksql> DROP EXTERNAL SOURCE my-postgres-jdbc;
```

**Sample Egress**

Data egress would work in a similar fashion:
```sql
ksql> CREATE EXTERNAL SINK my-elastic-sink WITH (sources='my-table1', ...);

ksql> SHOW EXTERNAL SINKS;

    name         |              connector                        |   topic prefix
-----------------|-----------------------------------------------|--------------------
my-elastic-sink  | io.confluent.connect.elasticsearch.ElasticsearchSinkConnector | elastic-sink

ksql> DESCRIBE my-elastic-sink;

name: my-elastic-sink
connector: io.confluent.connect.elasticsearch.ElasticsearchSinkConnector
topic-prefix: elstic-sink

   worder id     | task | status
-----------------|------|---------
10.200.7.69:8083 | 1    | RUNNING

name           | topic
--------------------------------------------
elasticsearch1 | elastic-sink-elasticsearch1 

ksql> DROP EXTERNAL SINK my-elastic-sink TERMINATE;
```

### Connect Improvements

This section is a record of some suggestions for Connect that would make it more KSQL-friendly:

- Connectors should have an API to expose all topics they create
- Connectors should have an API to map a topic to the schema that they create
- The JDBC connector should not require an SMT to populate the kafka KEY
- The Security model should work well with systems that create connectors "on behalf of" other users
- A more detailed error endpoint to help diagnose faulty connectors

### Advanced Topics

- **Error Handling**: it is difficult to understand whether a given connector is or is not working,
  and further difficult to debug why it is not working. This is out of scope for this KLIP and
  should be addressed in connect and then exposed externally. There is a limited set of synchronous 
  errors that can be easily identified (e.g. missing configurations) that can be bubbled up directly 
  to the user at the time when they submit the request.
- **Secret Management**: some connectors (e.g. JDBC connector) may require passing secrets to
  access the database. For the MVP, the security will naively rely on HTTPS and Connect - requiring
  users to specify their secrets in the SQL statement (e.g. `CREATE ... WITH(foo.username='me', 
  foo.password='secret')`). Although this is not ideal, these command will not be distributed onto
  the KSQL command topic. It is recommended that the Connect configuration topic, though, is locked
  down to prevent these secrets from leaking there. In a later iteration, KSQL can integrate more
  directly with Connect security features by passing tokens to some secret manager (e.g. Vault) and
  resolving them only server-side.

## Test plan

The test plan will involve writing a full demo application for the motivating use case, where the
user table is stored in Postgres. Beyond that, integration tests that spin up connect will be
added to ensure that compatibility does not degrade across releases.

## Documentation Updates

TODO

# Compatibility Implications

N/A

## Performance Implications

N/A

## Security Implications

See section on "Secret Management"

### CODE REVIEW

File: design-proposals/README.md [hunk 1]
```diff
@@ -47,4 +47,4 @@ Next KLIP number: **8**
 | [KLIP-3: Serialization of single Fields](klip-3-serialization-of-single-fields.md) | Approved       | N/A     |
 | [KLIP-4: Custom Type Registry](klip-4-custom-types.md)                             | Approved       | N/A     |
 | [KLIP-6: Execution Plans](klip-6-execution-plans.md)                               | Approved       | N/A     |
-| KLIP-7: Kafka Connect Integration                                                  | Proposal       | N/A     |
+| [KLIP-7: Kafka Connect Integration](klip-7-connect-integration.md)                 | Discussion     | N/A     |
```
Comment (big-andy-coates) [inline design-proposals/README.md]: Commit this index update in a separate PR so it can be merged now.

File: design-proposals/klip-7-connect-integration.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+- A single connector can have many topics, which makes modeling them as a single statement difficult
+- A connector can be modeled in different ways (e.g. stream/table) and connector lifecycle should
+  be independent with how KSQL interprets it
+
+In the future, there can exist shortcuts (similar to `C*AS`) to both create connectors and declare 
+corresponding streams/tables.
+
+There are a few more minor points of consideration:
+- All "DDL" commands that cause changes in connect will not be submitted to the command topic;
+  instead they will be issued on the REST server (this follows the pattern that all external-state
```
Comment (big-andy-coates) [inline design-proposals/klip-7-connect-integration.md]: An alternative to having a KSQL cluster know about a Connect cluster is to have the same logic in the CLI.  This has the benefit that we keep the KSQL server and Connect decoupled.

Aside from keeping the KSQL server code cleaner this also means a different users / teams can use different Connect clusters with the same KSQL cluster, i.e. the mapping changes from being:

> A KSQL cluster is configured to work with a connect cluster

To

> A KSQL CLI is configured to work with a KSQL cluster and a Connect cluster.

And we can have CLI commands to switch Connect cluster, as we do for KSQL cluster.

In effect, I'm proposing the integration point for Connect/KSQL should be in the CLI / UI, not in the KSQL server.

File: design-proposals/klip-7-connect-integration.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+-----------------|------|---------
+10.200.7.69:8083 | 0    | RUNNING
+
+ name      |     topic                     |  schema
+-----------|-------------------------------|--------------------------------------
+user-table | test-postgres-jdbc-user-table | id VARCHAR, user VARCHAR, age INTEGER
+page-table | test-postgres-jdbc-page-table | id VARCHAR, page VARCHAR, url VARCHAR
+```
+
+Then the user can create streams or tables from the underlying topics:
```
Comment (big-andy-coates) [inline design-proposals/klip-7-connect-integration.md]: Do we not what such automagically populated into the KSQL metastore?

File: design-proposals/klip-7-connect-integration.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+connector: io.confluent.connect.jdbc.SourceConnector
+topic-prefix: test-postgres-jdbc
+
+   worder id     | task | status
+-----------------|------|---------
+10.200.7.69:8083 | 0    | RUNNING
+
+ name      |     topic                     |  schema
+-----------|-------------------------------|--------------------------------------
+user-table | test-postgres-jdbc-user-table | id VARCHAR, user VARCHAR, age INTEGER
```
Comment (big-andy-coates) [inline design-proposals/klip-7-connect-integration.md]: Where is the schema coming from here? Does the Connect API have enough detail to provide this for non-Avro formats?

File: design-proposals/klip-7-connect-integration.md [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+terminate the connectors.
+```sql
+ksql> DROP EXTERNAL SOURCE my-postgres-jdbc TERMINATE;
+```
+
+**Sample Egress**
+
+Data egress would work in a similar fashion:
+```sql
+ksql> CREATE EXTERNAL SINK my-elastic-sink FROM my-table1 WITH (...);
```
Comment (big-andy-coates) [inline design-proposals/klip-7-connect-integration.md]: This syntax seems to assume a 1-2-1 mapping from topic to sink connector. Just like source connectors, I don't believe this is the case.

File: design-proposals/klip-7-connect-integration.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+elasticsearch1 | elastic-sink-elasticsearch1 
+
+ksql> DROP EXTERNAL SINK my-elastic-sink TERMINATE;
+```
+
+### Advanced Topics
+
+- **Error Handling**: it is difficult to understand whether a given connector is or is not working,
+  and further difficult to debug why it is not working. This is out of scope for this KLIP and
+  should be addressed in connect and then exposed externally. There is a class of synchronous errors
```
Comment (big-andy-coates) [inline design-proposals/klip-7-connect-integration.md]: Connect does expose status / errors etc though its API and we should be covering it as part of the MVP / KLIP.

There may be feedback we can give when making this integration and improvements we can make, but we should at least make initial use of what is there.

File: design-proposals/README.md (hunk 1 above)
Comment (agavra) [reply design-proposals/README.md]: I already did :) this is just adding a link

File: design-proposals/klip-7-connect-integration.md (hunk 2 above)
Comment (agavra) [reply design-proposals/klip-7-connect-integration.md]: I'm a little nervous about the portability of that approach, though I don't feel that strongly about it. What if I wanted to start my connectors as part of my sandbox deployment (e.g. headless mode)?

File: design-proposals/klip-7-connect-integration.md (hunk 3 above)
Comment (agavra) [reply design-proposals/klip-7-connect-integration.md]: yes we do! I think I explicitly called that out somewhere (as syntactic sugar), but I don't think that's part of an MVP.

File: design-proposals/klip-7-connect-integration.md (hunk 4 above)
Comment (agavra) [reply design-proposals/klip-7-connect-integration.md]: No it does not - this functionality would only work when it auto-registers with SchemaRegistry. For non-Avro formats, Connect would need to be extended to support exposing a schema API.

File: design-proposals/klip-7-connect-integration.md (hunk 5 above)
Comment (agavra) [reply design-proposals/klip-7-connect-integration.md]: This assumes 1-to-N, not 1-to-1. Is there a requirement for N-to-1?

File: design-proposals/klip-7-connect-integration.md (hunk 6 above)
Comment (agavra) [reply design-proposals/klip-7-connect-integration.md]: Definitely - see the `DESCRIBE <connector>` section: that will list out whether or not a certain connector is functioning or not, which is the basic MVP requirement. 

I'm not sure that we should expose more than this - KSQL's strength isn't in helping you visualize error information and debug connect. If we went any further then this it would likely cause a large maintenance burden whenever connect changes.

**Comment by agavra** [comment]
Thanks for the review @big-andy-coates! Most comments in-line

> I think it's imperative that you try and get something working end-to-end using the existing Connect APIs to find those details, i.e. try and injest two tables, where the PK has a different name, as tables into KSQL.

This is exactly what I did with a toy example and didn't run into anything blocking, though perhaps my example was too naive. I do think, though, that this KIP shouldn't address changes to individual connectors and the solution should be generic enough to work with any connector.

File: design-proposals/klip-7-connect-integration.md (hunk 5 above)
Comment (agavra) [reply design-proposals/klip-7-connect-integration.md]: Looks like connectors do support N-to-1. I'm not sure what the difference is here between spinning up 2 connectors (one for each topic) as opposed to one connector reading from two topics.

File: design-proposals/klip-7-connect-integration.md [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+--------------
+STREAM CREATED
+--------------
+```
+
+At this point, assuming the `pageviews_original` stream is already created, we can join the newly
+created `users_original` table with the `pageviews_original` stream. Finally, the user can drop and 
+terminate the connectors.
+```sql
+ksql> DROP EXTERNAL SOURCE my-postgres-jdbc TERMINATE;
```
Comment (vcrfxia) [inline design-proposals/klip-7-connect-integration.md]: What does the `TERMINATE` at the end of this statement signify? Will there be an option to drop an external source without terminating the connector?

Comment (agavra) [reply design-proposals/klip-7-connect-integration.md]: I was imagining you'd be able to remove the metadata for a connector without actually terminating the connector - though I suppose at second though I'm not sure why that would be especially useful.

File: design-proposals/klip-7-connect-integration.md (hunk 2 above)
Comment (big-andy-coates) [reply design-proposals/klip-7-connect-integration.md]: We can also have a rest API that knows how to deal with both, and forward statements to either the Connect RESTful API or the KSQL RESTful API appropriately.  What I think we should avoid is polluting the KSQL RESTFul API / Engine with Connect specific code.  We need to maintain good separation of concern / architectural boundaries.

File: design-proposals/klip-7-connect-integration.md (hunk 5 above)
Comment (big-andy-coates) [reply design-proposals/klip-7-connect-integration.md]: I would think multiplexing many topics through one connector is about minimising resource usage both in the Connect cluster and the external systems themselves.  Think about a source db that is being polled for changes. Much better to have one thing polling for 10 tables, rather than ten things polling independently and adding more load on the db.

The same is true of both source and sink connectors.

File: design-proposals/klip-7-connect-integration.md (hunk 6 above)
Comment (big-andy-coates) [reply design-proposals/klip-7-connect-integration.md]: I'm not we can say KSQL shouldn't be exposing more here. We are positioning this as an integrated solution, so we should be enabling users to setup these connectors purely through the KSQL APIs, not forcing them to use the Connect API to work out what is going on.  We should be surfacing as much information as possible from Connect to allow users to get things working.

**Comment by big-andy-coates** [comment]
> This is exactly what I did with a toy example and didn't run into anything blocking, though perhaps my example was too naive. I do think, though, that this KIP shouldn't address changes to individual connectors and the solution should be generic enough to work with any connector.

Agreed that we should work with any Connector.  What I'm alluding to here is performing a real work test of importing multiple tables with the JDBC source connector, where each table has a different primary key column name and ensuring the data is coming into KSQL with the key of the Kafka message populated.   Why?  Because this is what users will likely be doing and there are some complexities here that we should understand.  Before we announce this to the community we should understand what UX we're providing for such real world setups as that may influence what we consider to be part of the MVP, e.g. do we need some special code to just smooth out the wrinkles for the JDBC connector initially, or do we need to include Connect improvements, etc.  Obviously the Connect improvements are outside the scope of this KLIP, but they should be included in the scope of the MVP.

**Comment by agavra** [comment]
> Obviously the Connect improvements are outside the scope of this KLIP, but they should be included in the scope of the MVP.

I'm currently updating the KLIP with some information around "blessed" connectors that KSQL will implement to give seamless integration (such as the JDBC connector). As time goes on and we improve connect, we will hopefully obviate the need for these blessed connectors.

File: design-proposals/klip-7-connect-integration.md (hunk 5 above)
Comment (agavra) [reply design-proposals/klip-7-connect-integration.md]: This makes sense - I think we should suggest to connect an "Add configuration to existing connector" type of API that would allow us to run one connector with many logical configs, adding one at a time.

In the meantime, we'll just have the flexible configuration where the connectors accept whatever source in the WITH clause and I'll drop the FROM clause.

File: design-proposals/klip-7-connect-integration.md (hunk 2 above)
Comment (agavra) [reply design-proposals/klip-7-connect-integration.md]: I agree with this design goal and will keep it in mind during implementation. I have a preference toward not having the CLI communicate directly with another service (especially when things like endpoint discovery become problematic), but we should definitely keep these distinct and modular.

Comment (spena) [reply design-proposals/klip-7-connect-integration.md]: One good reason to integrate the Connect code in the CLI is around security. The user/password of the databases supplied won't go to the KSQL server which can be seen by the KSQL service principal (if we don't trust on the KSQL service principal of course), instead, they will go directly to Connect. However, a drawback is what if Connect is open for connections from the KSQL server only and not from any other client? That would not allow our CLI to communicate with Connect. It would be good to learn about how users would use Connect and KSQL and how they will configure it, right?

File: design-proposals/klip-7-connect-integration.md [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+**Sample Ingress**
+
+The first step is to create and start a Connector via the REST API (`POST /connectors`)
+```sql
+ksql> CREATE EXTERNAL SOURCE my-postgres-jdbc WITH ( \
+      connector.class="io.confluent.connect.jdbc.jdbcSourceConnector", \ 
+      tasks.max=1, \
+      connection.url=jdbc:postgresql://localhost:5432/my-db, \
+      ...);
```
Comment (spena) [inline design-proposals/klip-7-connect-integration.md]: Is the `connector.class` a required option for the `CREATE EXTERNAL SOURCE`? Also, I see that `DESCRIBE source_name` and `SHOW EXTERNAL SOURCES` has an output with headers that use `connector`. Also we have `GET /connectors`.

Are we planning of using the `CREATE EXTERNAL SOURCE` for other external sources other than connect at some point? Seems we are making this DDL command for Connect only, why not calling it `CREATE CONNECTOR`? Isn't more appropriate for it?

File: design-proposals/klip-7-connect-integration.md [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+Next, we provide tools to understand the status of the connector and corresponding topics hitting
+the `GET /connectors` and `GET /connectors/<name>` REST endpoints:
+```sql
+ksql> SHOW EXTERNAL SOURCES;
+
+    name         |              connector                        |   topic prefix
+-----------------|-----------------------------------------------|--------------------
+my-postgres-jdbc | io.confluent.connect.jdbc.JdbcSourceConnector | test-postgres-jdbc
+
+ksql> DESCRIBE my-postgres-jdbc;
```
Comment (spena) [inline design-proposals/klip-7-connect-integration.md]: Would this conflient with streams that have the same name? Let's say you have a `p1` connector for testing and a `p1` stream in KSQL. What would DESCRIBE describe?

File: design-proposals/klip-7-connect-integration.md [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+terminate the connectors.
+```sql
+ksql> DROP EXTERNAL SOURCE my-postgres-jdbc;
+```
+
+**Sample Egress**
+
+Data egress would work in a similar fashion:
+```sql
+ksql> CREATE EXTERNAL SINK my-elastic-sink WITH (sources='my-table1', ...);
```
Comment (spena) [inline design-proposals/klip-7-connect-integration.md]: Same question about the use of `CREATE CONNECTOR`. Btw, is there a difference for KSQL when creating a source connector and a sink connector? Doesn't Connecto differentiate from this by looking at the source or sink properties?

File: design-proposals/klip-7-connect-integration.md [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+
+   worder id     | task | status
+-----------------|------|---------
+10.200.7.69:8083 | 1    | RUNNING
+
+name           | topic
+--------------------------------------------
+elasticsearch1 | elastic-sink-elasticsearch1 
+
+ksql> DROP EXTERNAL SINK my-elastic-sink TERMINATE;
```
Comment (spena) [inline design-proposals/klip-7-connect-integration.md]: Isn't `CASCADE` a more standard name databases use?

Comment (agavra) [reply design-proposals/klip-7-connect-integration.md]: Oops haven't updated this - I'm going to remove the `TERMINATE` option entirely

File: design-proposals/klip-7-connect-integration.md (hunk 8 above)
Comment (agavra) [reply design-proposals/klip-7-connect-integration.md]: Sorry I'm not sure I get all of the questions you have there but I'll try to answer my best:

- `GET /connectors` is the call we'll be making _to_ connect
- For now and for the MVP only connect will be supported
- `CONNECTOR` doesn't differentiate between SOURCE/SINK, so between `CREATE SOURCE CONNECTOR` and `CREATE EXTERNAL SOURCE` I don't think there's much of a difference

File: design-proposals/klip-7-connect-integration.md (hunk 9 above)
Comment (agavra) [reply design-proposals/klip-7-connect-integration.md]: good point... I'm not sure how to address this issue. Suggestions for an alternative syntax?

File: design-proposals/klip-7-connect-integration.md (hunk 10 above)
Comment (agavra) [reply design-proposals/klip-7-connect-integration.md]: There is no difference, yet, for KSQL. The difference comes when we want to have more first class integration - e.g. we want to be able to do something like `CREATE EXTERNAL SINK foo FROM blah`. I think it's a little more extensible, but it's not that big of a point of contention for me.

**Comment by agavra** [comment]
Thanks for the review @spena! 

> What about the configuration in KSQL to access a Connect cluster? How would that be defined?

Via KsqlConfig (and the default will be embedded) they would pass in the URL for the connect cluster to use.

> Will KSQL support multiple Connect clusters?

For the MVP, no. Later there is no reason why we can't support multiple clusters but we would need to add syntax to differentiate which cluster we want to send things to.

File: design-proposals/klip-7-connect-integration.md (hunk 9 above)
Comment (spena) [reply design-proposals/klip-7-connect-integration.md]: `DESCRIBE CONNECTOR <connector-name>`?

Snowflake has a reserved word after `DESCRIBE` for each object type they support. See https://docs.snowflake.net/manuals/sql-reference/sql/desc.html

I see other databases do not have anything like that. But, their objects are part of their servers, and they can provide object names checks to avoid conflicts with other objects types (such as table vs function names). However, in our case, we will have a Connector that is part of an external service, so KSQL cannot do any naming conflict checks. I think `DESCRIBE CONNECTOR <name>` would help.

File: design-proposals/klip-7-connect-integration.md [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+--------------
+```
+
+*Note: as Connect manages to simplify the configuration, whether through templates or any other
+means, we will immediately be able to leverage this simplification in KSQL's WITH clause.*
+
+Next, we provide tools to understand the status of the connector and corresponding topics hitting
+the `GET /connectors` and `GET /connectors/<name>` REST endpoints:
+```sql
+ksql> SHOW CONNECTORS;
```
Comment (rodesai) [inline design-proposals/klip-7-connect-integration.md]: Would this show all connectors, or just those registered via KSQL? If the latter, how is that done?

Comment (rodesai) [inline design-proposals/klip-7-connect-integration.md]: Should this be `SHOW SOURCE CONNECTORS`? Or is the `SOURCE/SINK` optional?

File: design-proposals/klip-7-connect-integration.md [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+means that the API that we design will not be able to leverage any "embedded-ness" of Connect,
+nullifying most of the benefits of the second approach. 
+
+There are still, however, benefits to packaging connect and KSQL in the same JVM for a seamless
+experience for new users. Namely, new users will only need to download a single package, spin up
+a single process and maintain only a single application. There are some (solvable) concerns with
+this approach, but for the reason outlined above, the syntax will be developed with the assumption
+that Connect is running as a separate component.
+
+**The recommendation is to support 1 & 3, tabling 2 unless a need for it comes in the future**
```
Comment (rodesai) [inline design-proposals/klip-7-connect-integration.md]: Is the proposal for the first version just (1) or support both (1) and (3)?

File: design-proposals/klip-7-connect-integration.md (hunk 8 above)
Comment (big-andy-coates) [reply design-proposals/klip-7-connect-integration.md]: I think `CREATE SOURCE CONNECTOR` is preferable to `CREATE EXTERNAL SOURCE`.

1. 'source connector' is used in the Connect ecosystem and docs, so easier for people to relate prior experience or reading the docs to KSQL terminology
2. 'source' is currently used in KSQL to refer to a source _table_ or _stream_.  In this sense one might expect`EXTERNAL SOURCE` to be something very similar, but it isn't.
3. using `CREATE SOURCE CONNECTOR` leaves us free to use `CREATE EXTERNAL SOURCE` later something else, e.g. if we enhance Connect/KSQL so that we can ingest single tables and internally multiplex this into a single connector.

File: design-proposals/klip-7-connect-integration.md (hunk 9 above)
Comment (big-andy-coates) [reply design-proposals/klip-7-connect-integration.md]: +1 for having a discriminator after `DESCRIBE`.  

Though if we're going with `CREATE EXTERNAL SOURCE` and `CREATE EXTERNAL SINK` then it should be `DESCRIBE EXTERNAL SOURCE foo` and `DESCRIBE EXTERNAL SINK bar`.   Which is a bit of a mouthful.

If it's `CREATE SOURCE CONNECTOR` we could get away with `DESCRIBE CONNECTOR`, assuming no name clash is possible between source and sink connectors.

File: design-proposals/klip-7-connect-integration.md (hunk 12 above)
Comment (big-andy-coates) [reply design-proposals/klip-7-connect-integration.md]: It should show all IMHO. We should avoid tracking connector state in KSQL.

Comment (agavra) [reply design-proposals/klip-7-connect-integration.md]: I think having SOURCE/SINK optional is a good idea, my original intention was to show all.

File: design-proposals/klip-7-connect-integration.md (hunk 13 above)
Comment (agavra) [reply design-proposals/klip-7-connect-integration.md]: both 1&3

**Comment by agavra** [comment]
Merged because of this has come to equilibrium and progress is nearly done 😂
